airbyte-cdk/python/airbyte_cdk/entrypoint.py

```diff
@@ -160,8 +161,36 @@ def read(
         if self.source.check_config_against_spec:
             self.validate_connection(source_spec, config)
 
-        yield from self.source.read(self.logger, config, catalog, state)
-        yield from self._emit_queued_messages(self.source)
+        stream_message_counter: DefaultDict[HashableStreamDescriptor, int] = defaultdict(int)
+        for message in self.source.read(self.logger, config, catalog, state):
+            yield self.handle_record_counts(message, stream_message_counter)
+        for message in self._emit_queued_messages(self.source):
+            yield self.handle_record_counts(message, stream_message_counter)
+
+    @staticmethod
+    def handle_record_counts(message: AirbyteMessage, stream_message_count: DefaultDict[HashableStreamDescriptor, int]) -> AirbyteMessage:
+        if message.type == Type.RECORD:
+            stream_message_count[message_utils.get_stream_descriptor(message)] += 1
+
+        elif message.type == Type.STATE:
+            if not message.state.stream:
+                raise ValueError("State message was not in per-stream state format, which is required for record counts.")
```

Comment:
Idea: in the future, I would love to clarify potential error types / classes, so that we can distinguish between internal CDK logic errors (like this one), or source data value errors.

I see this as `AirbyteCDKError("State message was ... ")`

No change is needed in this PR — changing error classes is clearly a very different project.

Reply:
Is there a guarantee that `self.source.read` never returns a message that is a global state?

Reply:
Yep guarantee. Sources have never implemented global state as that was only for CDC db use cases. Maybe in the future but this is not something we've had any utility from.